datafusion-cli/src/cli_context.rs

```diff
@@ -15,19 +15,69 @@
 // specific language governing permissions and limitations
 // under the License.
 
-use std::sync::Arc;
+use std::num::NonZeroUsize;
+use std::sync::{
+    atomic::{AtomicBool, Ordering},
+    Arc, RwLock,
+};
 
 use datafusion::{
     dataframe::DataFrame,
     error::DataFusionError,
-    execution::{context::SessionState, TaskContext},
+    execution::{
+        context::SessionState,
+        memory_pool::{
+            MemoryConsumer, MemoryLimit, MemoryPool, MemoryReservation,
+            TrackConsumersPool, TrackedPool,
+        },
+        runtime_env::RuntimeEnvBuilder,
+        session_state::SessionStateBuilder,
+        TaskContext,
+    },
     logical_expr::LogicalPlan,
     prelude::SessionContext,
 };
 use object_store::ObjectStore;
 
 use crate::object_storage::{AwsOptions, GcpOptions};
 
+#[derive(Debug)]
```

Comment:
As I understand this code, I think it makes an `Arc<dyn MemoryPool>` implement the MemoryPool API directly

However since we can already copy an `Arc` I don't undertand  why this code is necessary so I played around with it.

Here is an alternate suggestion that I think is more general -- basically automatically derive MemoryPool for any `Arc<dyn MemoryPool>`:
- https://github.com/kosiew/datafusion/pull/30